File: bisect.hs

bisect :: Int -> [Int] -> Int
bisect a l = bisect_helper a l 0 (length l)
  where
    bisect_helper a l from to = if left < 2 then from
                                else if a < l !! half then bisect_helper a l from half
                                     else bisect_helper a l half to
      where
        half = div (from + to) 2
        left = to - from

bisect_test = do
  let l = [0 .. 10000]
  let e = bisect 10 l
  print $ "index is " ++ show e

main = do
  bisect_test

File: py/bisect.py

l = list(range(0, 100_000_000))
f = 0
t = len(l)
a = 5.5

def bisect(a, l : list, f=0, t=len(l)):
    if t - f < 2:
        return f
    h = (t + f) // 2
    if a < l[h]:
        return bisect(a, l, f, h)
    return bisect(a, l, h, t)

'''
while f + 1 < t:
    h = (f + t) // 2
    if a < l[h]:
        t = h
    else:
        f = h
'''

f = bisect(a, l)

print('index is ', f)

